java-checks/src/main/java/org/sonar/java/checks/spring/MissingPathVariableAnnotationCheck.java

```diff
@@ -0,0 +1,374 @@
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S6856")
+public class MissingPathVariableAnnotationCheck extends IssuableSubscriptionVisitor {
+  private static final String PATH_VARIABLE_ANNOTATION = "org.springframework.web.bind.annotation.PathVariable";
+  private static final String MAP = "java.util.Map";
+  private static final String MODEL_ATTRIBUTE_ANNOTATION = "org.springframework.web.bind.annotation.ModelAttribute";
+  private static final String REQUEST_MAPPING_ANNOTATION = "org.springframework.web.bind.annotation.RequestMapping";
+
+  private static final Set<String> MAPPING_ANNOTATIONS = Set.of(
+    REQUEST_MAPPING_ANNOTATION,
+    "org.springframework.web.bind.annotation.GetMapping",
+    "org.springframework.web.bind.annotation.PostMapping",
+    "org.springframework.web.bind.annotation.PutMapping",
+    "org.springframework.web.bind.annotation.DeleteMapping");
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree clazzTree = (ClassTree) tree;
+
+    List<MethodTree> methods = clazzTree.members().stream()
+      .filter(member -> member.is(Tree.Kind.METHOD))
+      .map(MethodTree.class::cast)
+      .toList();
+
+    // request @RequestMapping can be put on top of a class, path template inside it will affect all the class methods
+    var requestMappingArguments = clazzTree.symbol().metadata().valuesForAnnotation(REQUEST_MAPPING_ANNOTATION);
+    Set<String> requestMappingTemplateVariables = new HashSet<>();
+    if (requestMappingArguments != null) {
+      Set<String> templateVars;
+      try {
+        templateVars = templateVariablesFromMapping(requestMappingArguments);
+      } catch (DoNotReport ignored) {
+        return;
+      }
+
+      if (templateVars != null) {
+        requestMappingTemplateVariables = templateVars;
+      }
+    }
+
+    // we find path variable annotations on method annotated with @ModelAttribute and extract the name
+    Set<String> modelAttributeMethodParameter = new HashSet<>();
+    for (var method : methods) {
+      if (!method.symbol().metadata().isAnnotatedWith(MODEL_ATTRIBUTE_ANNOTATION)) {
+        continue;
+      }
+      for (var parameter : method.parameters()) {
+        SymbolMetadata metadata = parameter.symbol().metadata();
+        var arguments = metadata.valuesForAnnotation(PATH_VARIABLE_ANNOTATION);
+        if (arguments != null) {
+          modelAttributeMethodParameter.add(extractPathMethodParameters(parameter, arguments).value);
+        }
+      }
+    }
+
+    for (var method : methods) {
+      if (!method.symbol().metadata().isAnnotatedWith(MODEL_ATTRIBUTE_ANNOTATION)) {
+        try {
+          checkParametersAndPathTemplate(method, modelAttributeMethodParameter, requestMappingTemplateVariables);
+        } catch (DoNotReport ignored) {
+        }
+      }
+    }
+  }
```

Comment:
To make the visit method a little bit lighter, we can extract some of these for loops into separate methods. This one for instance could overload the `checkParametersAndPathTemplate` by taking a List of methods instead of a single one:

```suggestion
    checkParametersAndPathTemplate(methods, modelAttributeMethodParameter, requestMappingTemplateVariables);
  }

  private void checkParametersAndPathTemplate(List<MethodTree> methods, Set<String> modelAttributeMethodParameter, Set<String> requestMappingTemplateVariables) {
    for (var method : methods) {
      if (!method.symbol().metadata().isAnnotatedWith(MODEL_ATTRIBUTE_ANNOTATION)) {
        try {
          checkParametersAndPathTemplate(method, modelAttributeMethodParameter, requestMappingTemplateVariables);
        } catch (DoNotReport ignored) {
        }
      }
    }
  }
```